packages/databricks-vscode/src/vscode-objs/StateStorage.ts

```diff
@@ -1,73 +1,100 @@
-            value
-        );
-    }
+    "databricks.autocompletion.skipConfigure": withType<boolean>()({
+        location: "workspace",
+        defaultValue: false,
+    }),
 
-    get skipAutocompleteConfigure() {
-        return this.context.workspaceState.get(
-            "databricks.autocompletion.skipConfigure",
-            false
-        );
-    }
+    "databricks.fixedRandom": withType<number>()({
+        location: "global",
+        getter: (storage, value) => {
+            if (value === undefined) {
+                value = Math.random();
+                storage.set("databricks.fixedRandom", value);
+            }
+            return value;
+        },
+    }),
 
-    set skipAutocompleteConfigure(value: boolean) {
-        this.context.workspaceState.update(
-            "databricks.autocompletion.skipConfigure",
-            value
-        );
-    }
+    "databricks.fixedUUID": withType<string>()({
+        location: "workspace",
+        getter: (storage, value) => {
+            if (value === undefined) {
+                value = randomUUID();
+                storage.set("databricks.fixedUUID", value);
+            }
+            return value;
+        },
+    }),
 
-    get skippedEnvsForDbConnect() {
-        return this.context.globalState.get<string[]>(
-            "databricks.debugging.skipDbConnectInstallForEnvs",
-            []
-        );
-    }
+    "databricks.debugging.skipDbConnectInstallForEnvs": withType<string[]>()({
+        location: "global",
+        defaultValue: [],
+        setter: (storage, value) => {
+            if (value === undefined || value.length === 0) {
+                return undefined;
+            }
+            const currentEnvs: string[] = storage.get(
+                "databricks.debugging.skipDbConnectInstallForEnvs"
+            );
+            if (!currentEnvs.includes(value[0])) {
+                currentEnvs.push(value[0]);
+            }
+            return currentEnvs;
+        },
+    }),
 
-    skipDbConnectInstallForEnv(value: string) {
-        const currentEnvs = this.skippedEnvsForDbConnect;
-        if (!currentEnvs.includes(value)) {
-            currentEnvs.push(value);
-        }
-        this.context.globalState.update(
-            "databricks.debugging.skipDbConnectInstallForEnvs",
-            currentEnvs
-        );
-    }
+    "databricks.lastInstalledExtensionVersion": withType<string>()({
+        location: "workspace",
+        defaultValue: "0.0.0",
+    }),
+};
+
+type ValueType<K extends keyof typeof Keys> = (typeof Keys)[K]["_type"];
+type GetterReturnType<D extends KeyInfo<any>> = D extends {getter: infer G}
+    ? G extends (...args: any[]) => any
+        ? ReturnType<G>
+        : undefined
+    : undefined;
+
+type DefaultValue<K extends keyof typeof Keys> =
+    "defaultValue" extends keyof (typeof Keys)[K]
+        ? ValueType<K>
+        : GetterReturnType<(typeof Keys)[K]>;
+
+type KeyInfoWithType<V> = KeyInfo<V> & {
```

Comment:
What are the benefits of keeping it separate from the KeyInfo? Can those additional types just be defined on the KeyInfo, for the simplicity sake?

Reply:
KeyInfo holds actual concrete data. The type fields are only types. Defining them on keyinfo means someone can try setting them when adding a new fields to keys.